Update pull request template

hanzei opened the pull request:
#### Summary
This PR updates the pull request template. The goals I have in mind are:
1. Remove ambiguity. The old template at a lot of unclear phrases, like `Touches critical sections of the codebase`.
2. Remove arbitrary work for contributors. Contributors had some arbitrary work to do, like checking the checkboxes and removes the intro text.
3. Remove arbitrary work for PM's and core committers. We had so manually close the corresponding HW ticket once the PR is merged. We can get around this, if we use the `Fixes` keyword.

With this goals in mind I:
- Moved all info text into markdown comments.
- Added references to https://developers.mattermost.com
- Updates the checklist. I'm 3/5 on the first two times, but 0/5 on the other three. Feedback would be appreciated.
- Added a note about the `Fixes` keyword.

I've taken inspiration for the kubernetes project (https://github.com/kubernetes/kubernetes/blob/master/.github/PULL_REQUEST_TEMPLATE.md) and the atom project (https://github.com/atom/atom/blob/master/.github/PULL_REQUEST_TEMPLATE/feature_change.md).

If this PR gets merged, I will submit a similar on the other repos.

jasonblais commented on PULL_REQUEST_TEMPLATE.md:
```diff
@@ -1,19 +1,30 @@
-Please make sure you've read the [pull request](http://docs.mattermost.com/developer/contribution-guide.html#preparing-a-pull-request) section of our [code contribution guidelines](http://docs.mattermost.com/developer/contribution-guide.html).
+<!-- Thanks contribution a pull request! Here are some tips for you:
```
Propose `Thank you for contributing a pull request! Here are a few tips to help you:

jasonblais commented on PULL_REQUEST_TEMPLATE.md:
```diff
@@ -1,19 +1,30 @@
 #### Ticket Link
-[Please link the GitHub issue or Jira ticket this PR addresses.]
+<!--
+Please link the GitHub issue, if this PR addresses a Help Wanted ticket e.g.
```
Propose

```
If this pull request addresses a Help Wanted ticket, please link the relevant GitHub issue, e.g.

     Fixes https://github.com/mattermost/mattermost-server/issues/XXXXX

Otherwise, link the Jira ticket.
```

jasonblais commented on PULL_REQUEST_TEMPLATE.md:
```diff
@@ -1,19 +1,30 @@
-- [ ] Includes text changes and localization file ([.../i18n/en.json](https://github.com/mattermost/mattermost-server/blob/master/i18n/en.json)) updates
-- [ ] Touches critical sections of the codebase (auth, upgrade, etc.)
+<!--
+Please check every fields. Please remove unrelated fields.
```
Propose `Please review each item, and remove those not relevant for your pull request.`

jasonblais commented on PULL_REQUEST_TEMPLATE.md:
```diff
@@ -1,19 +1,30 @@
+<!--
+Please check every fields. Please remove unrelated fields.
+-->
+- Run test on the related changes locally
```
Do we have instructions on how to run these tests? E.g. running `make test`

hanzei replied:
There isn't a simple one liner for this. But we can link to https://developers.mattermost.com/contribute/server/developer-workflow/.

```suggestion
- Run test on the related changes locally <!-- See https://developers.mattermost.com/contribute/server/developer-workflow/ for an explanation how to do this -->
```

hanzei replied:
The problem is that `make test` runs _all_ test which takes some time and can even be to much for a weaker computer. Running just individual tests is usually recommended.

jasonblais replied:
Ah okay. In that case we could link to https://developers.mattermost.com/contribute/server/developer-workflow/#running-only-specific-server-unit-tests

```
Tested related changes <!-- See https://developers.mattermost.com/contribute/server/developer-workflow/#running-only-specific-server-unit-tests how to run specific unit tests. -->
```

jasonblais commented on PULL_REQUEST_TEMPLATE.md:
```diff
@@ -1,19 +1,30 @@
+- Run test on the related changes locally
+- Added or updated unit tests
+- Added REST API documentation (required for all new REST API endpoints)
+- Has enterprise changes (please link)
```
`Has` --> `Added` for consistency with other items.

Also `enterprise` --> `Enterprise`

jasonblais commented on PULL_REQUEST_TEMPLATE.md:
```diff
@@ -1,19 +1,30 @@
+- Added or updated unit tests
+- Added REST API documentation (required for all new REST API endpoints)
+- Has enterprise changes (please link)
+- Includes text changes and localization file ([.../i18n/en.json](https://github.com/mattermost/mattermost-server/blob/master/i18n/en.json)) updates
```
`Includes` --> `Included`

hanzei commented:
I would like to move `ISSUE_TEMPLATE.md` and `PULL_REQUEST_TEMPLATE.md` into `.github` to keep the repo more organised. Are you fine with this?

hanzei commented:
I don't think there are side effects. This is just a naming think. See https://help.github.com/articles/creating-a-pull-request-template-for-your-repository/ for the documentation.

jasonblais commented:
Then I don't see any problems with doing so.

lieut-data commented on .github/PULL_REQUEST_TEMPLATE.md:
```diff
@@ -0,0 +1,30 @@
+<!-- Thank you for contributing a pull request! Here are a few tips to help you:
```
Am I correct in guessing that Markdown comments (HTML comments, I guess?) automatically hides these 
blocks on display? Neat!

It will be interesting to see if this creates more or less problems for contributors in terms of leaving accidental comments. What would you say to commenting each line vs. block commenting, to make it easier to "remove" these without accidentally leaving a stray closing comment tag behind?

hanzei replied:
Yes that are HTML comments.

I don't see the need to remove the comments. Usually only the contributor edits the PR description. Hence all comments (expect for the ticket link one) can stay in my opinion.

lieut-data commented on .github/PULL_REQUEST_TEMPLATE.md:
```diff
@@ -0,0 +1,30 @@
+<!--
+If this pull request addresses a Help Wanted ticket, please link the relevant GitHub issue, e.g.
+
+  Fixes https://github.com/mattermost/mattermost-server/issues/XXXXX
```
nit: Does `Fixes: ` work and retain the desired functionality? Can one list multiple issues?

Separately, we had this kind of functionality enabled in JIRA for a while, but we had issues where a PR would close the ticket even though there were outstanding webapp and redux changes. Will we have that same problem here?

hanzei replied:
> nit: Does `Fixes: ` work and retain the desired functionality?

According to https://help.github.com/articles/closing-issues-using-keywords/ `Fixes:` works fine. Why do you wan the `:`?

>  Can one list multiple issues?

You need multiple `fixed` to close multiple issues, e.g. `Fixes #X, fixes #Y`.


> Separately, we had this kind of functionality enabled in JIRA for a while, but we had issues where a PR would close the ticket even though there were outstanding webapp and redux changes. Will we have that same problem here?

That's a good point. But there is a minority of contributions which target multiple repos.

But you'r right, this is a drawback. For me it's still worth to use `fixes`.

lieut-data commented on .github/PULL_REQUEST_TEMPLATE.md:
```diff
@@ -0,0 +1,30 @@
+<!--
+Please review each item, and remove those not relevant for your pull request
+-->
+- Tested related changes <!-- See https://developers.mattermost.com/contribute/server/developer-workflow/#running-only-specific-server-unit-tests how to run specific unit tests. -->
```
Intentionally omitting the `[ ]` for checkboxes?

hanzei replied:
Yes. The old checklist had some unnecessary steps IMO. You would check every item, remove the unrelated ones and then also check the boxes. IMO the last step is unneeded.

hanzei replied:
Added them again does feedback.

lieut-data commented on .github/PULL_REQUEST_TEMPLATE.md:
```diff
@@ -0,0 +1,30 @@
+Otherwise, link the Jira ticket.
+-->
+
+#### Checklist
```
@hanzei, would you consider sharing this with the team at large? This particular checklist predates me, so I don't have as much context into perhaps what went into the thinking here.

hanzei replied:
Sure, will share with in the Contributors channel.

hanzei replied:
Done https://community-daily.mattermost.com/core/pl/1ikmqhzw1ffsfepzyb4575bkih

hanzei commented:
Updated per feedback on the community server. Please review again.

lieut-data commented on .github/PULL_REQUEST_TEMPLATE.md:
```diff
@@ -0,0 +1,29 @@
+-->
+- [ ] Tested related changes <!-- See https://developers.mattermost.com/contribute/server/developer-workflow/#running-only-specific-server-unit-tests how to run specific unit tests. -->
+- [ ] Added or updated unit tests <!-- This should be done for every change, if possible -->
+- [ ] Added REST API documentation (required for all new REST API endpoints)
```
Could `(required for all new REST API endpoints)` go into a comment? (Since it's presumably meant for the PR author, and not the reviewers?)

lieut-data replied:
Also, is there documentation to which we could link that explains how to do this?

lieut-data commented on .github/PULL_REQUEST_TEMPLATE.md:
```diff
@@ -0,0 +1,29 @@
+<!--
+Please review each item and check the ones applicable for your pull request
+-->
+- [ ] Tested related changes <!-- See https://developers.mattermost.com/contribute/server/developer-workflow/#running-only-specific-server-unit-tests how to run specific unit tests. -->
```
I don't recall the specifics of the conversation on the developers channel, but I'm wondering if we need this line. If this particular documentation item is important enough to be called out here, I'm thinking we should just promote it somewhere on https://developers.mattermost.com/contribute/getting-started/contribution-checklist/ to make sure it gets read first.

It's also a bit confusing have a `Tested related changes` and `Added or updated unit tests`, since -- at the moment -- they amount to roughly the same things.

hanzei replied:
My thought was that I wanted to avoide the situation where someone submits a PR that doesn't even build. But maybe this a bit to much to add this item just for these few PR's.

lieut-data commented on .github/PULL_REQUEST_TEMPLATE.md:
```diff
@@ -0,0 +1,29 @@
+Please review each item and check the ones applicable for your pull request
+-->
+- [ ] Tested related changes <!-- See https://developers.mattermost.com/contribute/server/developer-workflow/#running-only-specific-server-unit-tests how to run specific unit tests. -->
+- [ ] Added or updated unit tests <!-- This should be done for every change, if possible -->
```
Do we need the parenthetical `This should be done for every change, if possible `? It seems like having shortened the list of required checkboxes, we should expect these to be done and no longer have to say that we expect them.

Also, what about shortening `Added or updated` to just:
```suggestion
- [ ] Updated unit tests
```

hanzei replied:
I see your point. But at the end, what is the whole point of this check list item anyway? Like, what does it change for the reviewers if this is checked or not? They will request unit test anyway if then think this is required.

lieut-data commented on .github/PULL_REQUEST_TEMPLATE.md:
```diff
@@ -0,0 +1,29 @@
+- [ ] Tested related changes <!-- See https://developers.mattermost.com/contribute/server/developer-workflow/#running-only-specific-server-unit-tests how to run specific unit tests. -->
+- [ ] Added or updated unit tests <!-- This should be done for every change, if possible -->
+- [ ] Added REST API documentation (required for all new REST API endpoints)
+- [ ] Included text changes and localization file ([.../i18n/en.json](https://github.com/mattermost/mattermost-server/blob/master/i18n/en.json)) updates
```
Could we simplify this to just say:
```suggestion
- [ ] Updated localizations ([.../i18n/en.json](https://github.com/mattermost/mattermost-server/blob/master/i18n/en.json))
```

hanzei commented:
I've been thinking a bit more about this and I start to question if the checklist is even needed. We shrinked it down to two items. Which are:
1. Updated unit tests
2. Added REST API documentation
3. Updated localizations ([.../i18n/en.json](https://github.com/mattermost/mattermost-server/blob/master/i18n/en.json))

As said above, I don't see the need for 1.

2. Is more like a todo later item. Maybe we can add a label for this, like `Docs: Needed`, e.g. `API-Docs: Needed`. 2/5

3. seams superfluous to me, too. A reviewer will check for this anyway.

jasonblais commented:
Seems like all those items are in https://developers.mattermost.com/contribute/getting-started/contribution-checklist/ except API docs. Maybe we add the API docs there, and link to that contribution checklist in the PR template?

0/5 on approach, but I do think there should be some reference for a "quick contribution checklist" especially for those who are new contributors.

Jesse might have insight on how much the checklist has helped core committers when submitting PRs.

lieut-data commented:
I agree with the general direction of pointing contributors at  https://developers.mattermost.com/contribute/getting-started/contribution-checklist/. While I did benefit from the checklist when starting, I /suspect/ the link to this checklist would be more valuable in general since it can go into detail on expectations.

I also find the current checklist cumbersome and would love to see it go away altogether!

One other piece of feedback: might we consider eliminating the `#### Summary` header altogether? When I submit a PR with a single commit, that commit's description ends up at the top of the PR description automatically, but in addition to removing all the superfluous checkboxes, I always have to move that description down below the `#### Summary` header. It would be neat for a certain class of PRs to be "ready" without any additional editing required.

hanzei commented:
Head up: Jesse and have a call scheduled on Monday to discus the final points. I will update and re-request reviews after this.

hanzei commented:
@jwilander This is ready for your review. Please take a look.
